### ISSUE TITLE
wire code gen extensions

### ISSUE THREAD

**Issue by strangemonad** [square/wire#1867]
I'm not really sure how to name issue - more of a general question. I can open something more specific if this leads anywhere. I didn't see any similar issue, sorry if I missed something.

There's a specific use case I'm trying to satisfy and a potentially more general way of accomplishing that solution that would fit other needs. I don't think I'm trying to propose something as general as the protoc plugin model.

Very concretely, I'm curious about trying to leverage validation annotations like the ones in https://github.com/envoyproxy/protoc-gen-validate but rather than generate a `Validator#validate(Message m)` for each message type, I'd like to have the generated object apply its own validations and throw an exception when trying to create an invalid instance. eg. adding `require` checks to a kotlin init block or very similar to how Immutables Java does precondition checks https://immutables.github.io/immutable.html#precondition-check-method

This same mechanism could possibly extend to supporting field annotations like redacted to exclude the field from `toString` generation?

Maybe doing domain modeling like this in protos is just a foolish idea?

**Comment by JakeWharton** [comment]
As you probably know, Square has a bunch of options for post-deserialization validation which are not honored by Wire. And it comes up a few times a year. I know @efirestone also has/had ambitions in this space. Unfortunately, unlike our other libraries (Moshi/Gson/Retrofit) we don't have a central registry in which adapters are looked up and can be wrapped–everything is statically linked.

The closest thing we have is the entirely-undocumented custom adapter feature. It's so undocumented I don't even remember if that is the name... and I don't know if it allows you to delegate to the real adapter for deserialization and then perform validation. But perhaps we can adjust it so that it's possible? That's a @swankjesse production.

**Comment by oldergod** [comment]
Btw custom adapters are only supported in Java.

**Comment by swankjesse** [comment]
I've got an idea for how to make this work. In the latest Moshi we added tags to JsonReader and JsonWriter. They're convenient little places to share state between adapters and the calling application.

For Wire, we could put something on ProtoReader and ProtoWriter, perhaps a per-object callback invoked N times for the N objects in a tree.

Could range from basic:

```
interface ProtoCallback {
  fun beforeWrite(adapter: ProtoAdapter<T>, value: T): T
  fun afterRead(adapter: ProtoAdapter<T>, value: T): T
}
```

Or more interceptorey:

```
interface ProtoInterceptor {
  fun write(writer: ProtoWriter, adapter: ProtoAdapter<T>, tag: Int, value: T)
  fun read(reader: ProtoReader, adapter: ProtoAdapter<T>, tag: Int): T
}
```

Some interesting actions to take:
 * Crash on invalid values (in or out)
 * Log invalid values
 * Rewrite values
 * Omit obsolete values

Doing a JsonReader-style path here would also be quite powerful.

**Comment by JakeWharton** [comment]
Could also switch to instance-linked adapters instead of a statically-linked ones, if you feel like a major change.

**Comment by swankjesse** [comment]
Instance-linked adapters would be more capable! But it’s a very big of a change. Migrating methods like `Message.encode()` scares me on that.

**Comment by JakeWharton** [comment]
Yep, they'd need removed. It would fully decouple the model objects from serialization such that serializing to other formats like text or json would be easier. Although there's an additional burden on needing to register the generated adapters.

**Comment by strangemonad** [comment]
I think I'm thinking more about a mechanism that might hook into the proto object instance instantiation. I'd like to be able to also handle the case where programmatically creating an instance can check the validators (e.g. as the builder calls the constructor) without necessarily needing any encoding/decoding. 

Once interesting note here. [Java Immutables](https://immutables.github.io/)  handles both validation and normalization with a similar mechanism. It might be that this could also provide a general way to register well known type mappings (e.g. Instant).
